grove/alpha/simon/simon.py

```diff
@@ -13,441 +13,325 @@
+        self.ancillas = None
+        self.simon_circuit = None
+        self.oracle_circuit = None
+        self._dict_of_linearly_indep_bit_vectors = {}
+        self.mask = None
+        self.bit_map = None
+        self.classical_register = None
+
+    def _construct_simon_circuit(self):
+        """
+        Implementation of the quantum portion of Simon's Algorithm.
+
+        Given a list of input qubits,
+        all initially in the :math:`\\vert 0\\rangle` state,
+        create a program that applies the Hadamard-Walsh transform the qubits
+        before and after going through the oracle.
+
+        :param Program oracle: Program representing unitary application of function
+        :param list(int) qubits: List of qubits that enter as the input
+                            :math:`\\vert x \\rangle`.
+        :return: A program corresponding to the desired instance of
+                 Simon's Algorithm.
+        :rtype: Program
+        """
+        p = pq.Program()
+
+        oracle_name = "SIMON_ORACLE"
+        p.defgate(oracle_name, self.unitary_function_mapping)
+
+        p.inst([H(i) for i in self.log_qubits])
+        p.inst(tuple([oracle_name] + sorted(self._qubits, reverse=True)))
+        p.inst([H(i) for i in self.log_qubits])
+        return p
+
+    def _init_attr(self, bitstring_map):
+        """
+        Acts instead of __init__ method to instantiate the necessary Simon Object state.
+
+        :param bitstring_map: truth-table of the input bitstring map in dictionary format
+        :return: None
+        """
+        self.bit_map = bitstring_map
+        self.n_qubits = len(list(bitstring_map.keys())[0])
+        self.n_ancillas = self.n_qubits
+        self._qubits = list(range(self.n_qubits + self.n_ancillas))
+        self.log_qubits = self._qubits[:self.n_qubits]
+        self.ancillas = self._qubits[self.n_qubits:]
+        self.classical_register = np.asarray(list(range(self.n_qubits + self.n_ancillas)))
+        self.unitary_function_mapping, _ = self._compute_unitary_oracle_matrix(bitstring_map)
+        self.simon_circuit = self._construct_simon_circuit()
+        self._dict_of_linearly_indep_bit_vectors = {}
+        self.mask = None
+        self.classical_register = np.asarray(list(range(self.n_qubits + self.n_ancillas)))
+
+    @staticmethod
+    def _compute_unitary_oracle_matrix(bitstring_map):
+        """
+        Computes the unitary matrix that encodes the orcale function for Simon's algorithm
+
+        :param bitstring_map: truth-table of the input bitstring map in dictionary format
+        :return: a dense matrix containing the permutation of the bit strings and a dictionary
+        containing the indices of the non-zero elements of the computed permutation matrix as
+        key-value-pairs
+        :rtype: Tuple[2darray, Dict[String, String]]
+        """
+        n_bits = len(list(bitstring_map.keys())[0])
+        ufunc = np.zeros(shape=(2 ** (2 * n_bits), 2 ** (2 * n_bits)))
```

Comment:
can you make a variable `num_bits = 2 * n_bits`, maybe? And example why there is a factor of two with a comment? This code is pretty mathy and I think it could use comments.

Reply:
this would look really weird in the code and be more obfuscating than helpful. I will include a comment though.